### ISSUE TITLE
Can instagram-scraper get the highest video quality of posts? (720p upwards) that seem to only be accessible on mobile

### ISSUE THREAD

**Issue by github-userx** [rarcega/instagram-scraper#436]
Or can this only be achieved through the API? I haven’t found any tools that can get HD or fullHD video of posts

**Comment by itsdante** [comment]
Sadly no, I tried with the instagram graph api, there's no HD video that they provide, it's the same with web crawling, there is no HD section available, there is an image preview tho. Hope this helps

**Comment by github-userx** [comment]
With the private API apparently we can get higher resolution video than with scraper or instaloader, but seems like still no one has written a tool for This, see:

https://github.com/notcammy/PyInstaStories/issues/14#issuecomment-526690436 <https://github.com/notcammy/PyInstaStories/issues/14#issuecomment-526690436>

**Comment by VasiliPupkin256** [comment]
@github-userx what exact API call you are discussing there? This app is now able to download videos of 640x1138 for stories, and is using graphgl API calls of a form https://www.instagram.com/p/B1cR5PHI4JA/?__a=1 to retrieve media. There is `dash_info` parameter in there with a link to a higher resolution mp4 file but we don't use it. It is segmented somehow, what other tool is able to combine those video segments and process all the possible parameters of this dash entity? I have also looked into https://github.com/ytdl-org/youtube-dl project which is known for its ability to decode such segmented videos but it also wasn't able to process this https://www.instagram.com/p/B1cR5PHI4JA/ url correctly

**Comment by github-userx** [comment]
the video with the shortcode BqPch1Yg_P9 - instaloader (and Instagram-scraper?) only gets the 270p version of it when there does exist a 360p Version according to the json output

**Comment by VasiliPupkin256** [comment]
@github-userx it is inside dash manifest, video is separated from audio and may also be split into segments, we can't just download it without further processing. Do you know any open source tool that has code to parse this manifest and remux video or at least a description of its content?

**Comment by github-userx** [comment]
ffmpeg? Can almost do anything.

streamlink does exist as well.

**Comment by VasiliPupkin256** [comment]
@github-userx it is not a problem to combine a video the question is how those segments are coded inside this manifest. How are you downloading it? I can only find this vague trial to understand its structure here mgp25/Instagram-API#1683

**Comment by github-userx** [comment]
Apologies, I’m probably too uneducated (not a programmer) to understand this even in the least :/

**Comment by VasiliPupkin256** [comment]
@github-userx what API call you have discussed with @notcammy in the issue you linked? How are you downloading this higher quality videos? How did you know that there is a higher quality video for this shortcode? Do you have any info other than the magic function in PyInstaStories?

**Comment by github-userx** [comment]
I only know the resolution thing because of @notcammy‘s JSON output:

https://pastebin.com/emRKqnxZ <https://pastebin.com/emRKqnxZ>

**Comment by notcammy** [comment]
> How are you downloading this higher quality video?

The MPD has separate file for audio and (different versions of) video. You can parse manifest with https://docs.python.org/2/library/xml.dom.minidom.html, download video and audio track as-is with Requests and merge them with ffmpeg, like I do with story videos. They don't seem to be split into segments.

> What API call you have discussed with @notcammy in the issue you linked? 

The app endpoint is defined here in the library I use: https://github.com/ping/instagram_private_api/blob/master/instagram_private_api/endpoints/media.py#L15
The headers are obviously different from web API so implementing this (taking authentication, cookies, device info etc. into account) will probably require considerable research and effort

**Comment by VasiliPupkin256** [comment]
> You can parse manifest with xml.dom.minidom

Parsing xml is not a problem, the question is what schema is used for this manifest xml, what is the meaning of all those tags. What is SegmentBase, indexRange, Initialization range, all this. As I understand you've reverse engineered its meaning from examples. You have just extracted the first `BaseURL` from the first `Representation` from the first `AdaptationSet` from the first `Period` assuming that there can't be any other segments and periods and the first `Representation` always has the highest resolution (i.e. `width` parameter) among others. Right? Does it always worked with no bugs and counterexamples and the whole MDP parsing logic is here notcammy/PyInstaStories/blob/master/pyinstastories.py#L186-L200

**Comment by notcammy** [comment]
>You have just extracted the first `BaseURL` from the first `Representation` from the first `AdaptationSet` from the first `Period` assuming that there can't be any other segments and periods and the first `Representation` always has the highest resolution (i.e. `width` parameter) among others. Right?

Right. For stories, the schema seems always pretty much identical. I can only assume the same is true for videos (but this of course ought to be properly researched first for a large project like this).

From my findings, highest resolution is always the first Representation in the first AdaptionSet. There is only one audio track in the second AdaptionSet (when audio is present), so that's why the parsing logic is as basic as it is.

I didn't bother finding out what the tags mean and how they are used as simply downloading and merging the files have worked for hundreds of files so far without problem in my experience. So far they don't seem very relevant for this use case.

**Comment by github-userx** [comment]
@notcammy I’m pretty confused (as I am quite the noob).

Is there a way right now that I can download the 360p version (instead of the 270p version) of this video: BqPch1Yg_P9 ?

If so, I would appreciate a short breakdown of the steps a lot. I can use the commandline, ffmpeg etc. but I’m just a bit confused with everything being discussed here and in the other thread.

**Comment by notcammy** [comment]
>using graphgl API calls of a form https://www.instagram.com/p/B1cR5PHI4JA/?__a=1 to retrieve media. There is `dash_info` parameter in there with a link to a higher resolution mp4 file but we don't use it.

I missed this while writing my earlier posts. If it's in graphql api then implementation should be easy and would only require extracting the video and audio file, then merging these files with ffmpeg.

@github-userx You can't bulk download higher quality videos until the maintainers of this project decide to implement what is described above.

**Comment by github-userx** [comment]
Thanks a lot! And single videos?

**Comment by notcammy** [comment]
@github-userx https://gist.github.com/notcammy/7c32385246cd0fa2f83b34aa6c6988e8

Usage: 
`python3 script.py --username johndoe --password grapefruit --download BqPch1Yg_P9`

You may omit username and password arguments when credentials.json file is generated. For login issues see https://github.com/notcammy/PyInstaLive/issues/57

You need ffmpeg and [this dependency](https://github.com/ping/instagram_private_api) installed. May or may not work with multiple videos in one post, and won't download anything at all if there is no dash manifest available.

**Comment by VasiliPupkin256** [comment]
Ok. Apparently this manifest standard is a part of MPEG DASH specification ISO/IEC 23009-1, ffmpeg can even process this MPD xml file directly. A more or less complete parser code on python can be reused from youtube-dl project for example.

**Comment by github-userx** [comment]
@notcammy what would be the useragent showing up in instagram activity log when using your instagram high-quality-video download script?

Is it just „Linux“ or some mobile device name?

**Comment by nvmnghia** [comment]
Somehow [`gallery-dl`][1] can squeeze a 720p story, but as of today Instagram Scraper can only do 640x1138.

  [1]: https://github.com/mikf/gallery-dl
